agent/consul/auto_config_endpoint.go

```diff
@@ -84,87 +79,77 @@ func (a *jwtAuthorizer) Authorize(req *agentpb.AutoConfigRequest) (AutoConfigOpt
 	}, nil
 }
 
-// Cluster endpoint is used for cluster configuration operations
-type Cluster struct {
-	srv *Server
+type AutoConfigDelegate interface {
+	GetConfig() *Config
+	CreateACLToken(template *structs.ACLToken) (*structs.ACLToken, error)
+	DatacenterJoinAddresses(segment string) ([]string, error)
+	TLSConfigurator() *tlsutil.Configurator
+	ForwardRPC(method string, info structs.RPCInfo, args, reply interface{}) (bool, error)
```

Comment:
What do you think of something like `AutoConfigBackend` for the name of this interface? Delegate is kind of implied by it being an interface.

Reply:
That sounds fine to me and I will make the change.

What do you think about the interface? In particular I was debating about whether this should all go into a single interface or multiple as well as whether the Config and TLS Configurator should just be provided to the AutoConfig RPC struct instead of retrieved via the interface.

Reply:
The interface looks reasonable to me. I started digging into how it was used a bit, but I ran out of time.

For config reload, an idea I have been playing with would be to have all the components that need to reload implement an interface. I'm not sure exactly what that needs to look like yet, but I think it might be something like this:

```go
type Reloadable interface {
    ReloadStart() (done func())
    UpdateConfig(config.Config)
```

Any component that needs to pause and lock all operations can do so with `ReloadStart` (and unlock with `done`), and everything would accept the new config from `UpdateConfig()`.

The agent would store a `[]Reloadable` and call all the `ReloadStart()` first, then all the `UpdateConfig()`, and finally all the `done()` in reverse order.

To get around the import cycle with `config` `UpdateConfig` might take a closure instead of a config struct, I'm not sure.

Reply:
So I made a few changes since your review.

I got rid of the `GetConfig` and `TLSConfigurator` methods in the interface. Instead those concrete types are passed into a constructor. I also noticed that the only fields on the Config that were used were not runtime reloadable so I didn't bother implementing the reloadability of the configuration (the TLS CA certs are runtime reloadable but the `Agent` handles updating the shared tls configurator)

As for whats left in the interface, the 3 functions only relation is that they are needed by this RPC endpoint. With another soon to be opened PR I will be adding a couple more related to signing certificates and getting the latest connect CA roots. Still there isn't a strong relationship between these methods. I could have split this into multiple interfaces but that seems a little excessive for now.